node/libs/protobuf_build/src/lib.rs

```diff
@@ -140,39 +145,93 @@ impl Config {
 
     /// Generates implementation of `prost_reflect::ReflectMessage` for a rust type generated
     /// from a message of the given `proto_name`.
-    fn reflect_impl(&self, proto_name: &ProtoName) -> anyhow::Result<syn::ItemImpl> {
+    fn reflect_impl(&self, proto_name: &ProtoName) -> anyhow::Result<syn::Item> {
         let rust_name = proto_name
             .relative_to(&self.proto_root.to_name().context("invalid proto_root")?)
             .unwrap()
             .to_rust_type()?;
         let proto_name = proto_name.to_string();
         let this = self.this_crate();
         Ok(syn::parse_quote! {
-            impl #this::prost_reflect::ReflectMessage for #rust_name {
-                fn descriptor(&self) -> #this::prost_reflect::MessageDescriptor {
-                    static INIT: #this::Lazy<#this::prost_reflect::MessageDescriptor> = #this::Lazy::new(|| {
-                        DESCRIPTOR.get_message_by_name(#proto_name).unwrap()
-                    });
-                    INIT.clone()
-                }
-            }
+            #this::impl_reflect_message!(#rust_name, &DESCRIPTOR, #proto_name);
         })
     }
 
-    /// Generates rust code from the proto files according to the config.
-    pub fn generate(&self) -> anyhow::Result<()> {
+    /// Validates this configuration.
+    fn validate(&self) -> anyhow::Result<()> {
+        /// Flag set to `true` if a public compilation target was encountered in a build script.
+        static HAS_PUBLIC_TARGET: AtomicBool = AtomicBool::new(false);
```

Comment:
Would it be hard to lift this constraint?

Reply:
Theoretically, it can be lifted by changing the exported variable name (`manifest` rn) depending on the compilation target; but I didn't come up with a good way to ID compilation targets and passing these IDs to dependent crates. IMO, restricting a crate to at most one public target makes sense – it makes sure that all public generated code is in a single place. Do you have use cases for multiple public targets in mind?

Reply:
For example roles crate could have a separate protobuf targets for validator and node roles. I don't see any large benefits though, so it can wait.